keps/sig-network/4762-allow-arbitrary-fqdn-as-pod-hostname/README.md

```diff
@@ -0,0 +1,554 @@
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposal allows users to set arbitrary Fully Qualified Domain Name (FQDN) as the hostname of a pod, introduces a new field `actualPodHostname` for the podSpec, which, if set, will always be respected by the Kubelet (otherwise it will fall back to legacy behavior), and no longer cares about the `hostname` as well as the `setHostnameAsFQDN` values.
```

Comment:
a new field?

this will make this more complicated than it is today, no?

why is not an alternative to allow to relax the validation of the current `pod.spec.hostname` field to be able to set a fqdn there?

Reply:
Yes, adding it will make everything more complex, but now we will create DNS records based on the hostname field in podSpec. If we set the FQDN in the hostname, it is undoubtedly a disruption to the existing behavior.
ref: https://github.com/kubernetes/kubernetes/blob/e1aa8197eddd277fb82be0b8bbc2ba2b4ca67af7/pkg/controller/endpoint/endpoints_controller.go#L439-L441

Reply:
why is disruptive? are those controllers not copying the field verbatim?

Reply:
When we set the `hostname` field in the `podSpec` to an FQDN, such as `www.google.com`, and also set the `subdomain` field, we end up with an FQDN hostname that appends the `cluster-suffix`, resulting in `www.google.com.subdomain.default.svc.cluster.local`. Additionally, we get a DNS record for `www.google.com.subdomain.default.svc.cluster.local`. I believe this does not align with our current design patterns, and it also can't meet our current needs.

Of course, we can also state that when the `hostname` in the `podSpec` is set to an FQDN, we can ignore the `subdomain` field. I will add this to the Alternatives section.

Reply:
... or to keep the current `hostname` and be able to  override or omit the `default.svc.cluster.local` part , right ? :)

Reply:
yes, added.